cpp/src/strings/regex/regcomp.cpp

```diff
@@ -714,97 +741,69 @@ class regex_compiler {
-      if (in[i].t != COUNTED && in[i].t != COUNTED_LAZY) {
-        out.push_back(in[i]);
-        if (in[i].t == LBRA || in[i].t == LBRA_NC) {
-          lbra_stack.push_back(i);
-          rep_start = -1;
-        } else if (in[i].t == RBRA) {
-          rep_start = lbra_stack[lbra_stack.size() - 1];
-          lbra_stack.pop_back();
-        } else if ((in[i].t & 0300) != OPERATOR_MASK) {
-          rep_start = i;
+    std::vector<regex_parser::Item> out;
+    std::stack<int> lbra_stack;
+    auto repeat_start_index = -1;
+
+    for (std::size_t index = 0; index < in.size(); index++) {
+      regex_parser::Item item = in[index];
+
+      if (item.t != COUNTED && item.t != COUNTED_LAZY) {
+        out.push_back(item);
+        if (item.t == LBRA || item.t == LBRA_NC) {
+          lbra_stack.push(index);
+          repeat_start_index = -1;
+        } else if (item.t == RBRA) {
+          repeat_start_index = lbra_stack.top();
+          lbra_stack.pop();
+        } else if ((item.t & ITEM_MASK) != OPERATOR_MASK) {
+          repeat_start_index = index;
         }
       } else {
-        if (rep_start < 0)  // broken regex
-          return;
+        // item is of type COUNTED or COUNTED_LAZY
+        // here we repeat the previous item(s) based on the count range in item
 
-        regex_parser::Item item = in[i];
-        if (item.d.yycount.n <= 0) {
-          // need to erase
-          for (std::size_t j = 0; j < i - rep_start; j++)
-            out.pop_back();
-        } else {
-          // repeat
-          for (int j = 1; j < item.d.yycount.n; j++)
-            for (std::size_t k = rep_start; k < i; k++)
-              out.push_back(in[k]);
+        CUDF_EXPECTS(repeat_start_index >= 0, "regex: invalid counted quantifier location");
+
+        // range of affected item(s) to repeat
+        auto const begin = in.begin() + repeat_start_index;
+        auto const end   = in.begin() + index;
+        // count range values
+        auto const n = item.d.yycount.n;  // minimum count
+        auto const m = item.d.yycount.m;  // maximum count
+
+        // zero-repeat edge-case: need to erase the previous items
+        if (n <= 0) { out.erase(out.end() - (index - repeat_start_index), out.end()); }
+
+        // minimum repeats (n)
+        for (int j = 1; j < n; j++) {
+          out.insert(out.end(), begin, end);
         }
 
-        // optional repeats
-        if (item.d.yycount.m >= 0) {
-          for (int j = item.d.yycount.n; j < item.d.yycount.m; j++) {
-            regex_parser::Item o_item;
-            o_item.t    = LBRA_NC;
-            o_item.d.yy = 0;
-            out.push_back(o_item);
-            for (std::size_t k = rep_start; k < i; k++)
-              out.push_back(in[k]);
+        // optional maximum repeats (m)
+        if (m >= 0) {
+          for (int j = n; j < m; j++) {
+            out.push_back(regex_parser::Item{LBRA_NC, 0});
+            out.insert(out.end(), begin, end);
```

Comment:
Similar to above, should be call `reserve` first?

Reply:
Thanks. There are some boundary cases that would require extra if-logic which I think would detract from the flow here. There is no performance issue here that requires `reserve()` so I'd rather keep the code as simple as possible.
This did get me thinking that I should add an assert here to ensure `n >= 0` since we are far removed from the logic that sets that value.